daisy_workflows/image_test/linux_common/utils.py

```diff
@@ -99,10 +185,185 @@ def RunTest(test_func):
+          body=md_obj)
+    request.execute()
+
+  def ExtractKeyItem(self, md_key, level):
+    """Extract a given key value from the metadata.
+
+    Args:
+      md_key: string, the key of the metadata value to be searched.
+      level: enum, INSTANCE_LEVEL or PROJECT_LEVEL to fetch the metadata.
+
+    Returns:
+      md_item: dict, in the format {'key', md_key, 'value', md_value}.
+      None: if md_key was not found.
+    """
+    for md_item in self.md_items[level]:
+        if md_item['key'] == md_key:
+            return md_item
+
+  def Define(self, md_key, md_value, level, store=True):
+    """Add or update a metadata key with a new value in a given level.
+
+    Args:
+      md_key: string, the key of the metadata.
+      md_value: string, value to be added or updated.
+      level: enum, INSTANCE_LEVEL or PROJECT_LEVEL to fetch the metadata.
+      store: bool, if True, saves metadata to GCE server.
+    """
+    if not level:
+        level = self.INSTANCE_LEVEL
+    md_item = self.ExtractKeyItem(md_key, level)
+    if md_item and md_value is None:
+      self.md_items[level].remove(md_item)
+    elif not md_item:
+      md_item = {'key': md_key, 'value': md_value}
+      self.md_items[level].append(md_item)
+    else:
+      md_item['value'] = md_value
+    if store:
+        self.StoreMetadata(level)
+
+  def AddSshKey(self, md_key, level=None, store=True):
+    """Generate and add an ssh key to the metadata
+
+    Args:
+      md_key: string, SSH_KEYS or SSHKEYS_LEGACY, defines where to add the key.
+      level: enum, INSTANCE_LEVEL (default) or PROJECT_LEVEL to fetch the
+          metadata.
+      store: bool, if True, saves metadata to GCE server.
+
+    Returns:
+      key_name: string, the name of the file with the generated private key.
+    """
+    if not level:
+        level = self.INSTANCE_LEVEL
+    key, key_name = GenSshKey(self.ssh_user)
+    md_item = self.ExtractKeyItem(md_key, level)
+    if not md_item:
+      md_item = {'key': md_key, 'value': key}
+      self.md_items[level].append(md_item)
+    else:
+      md_item['value'] = '\n'.join([md_item['value'], key])
+    if store:
+        self.StoreMetadata(level)
+    return key_name
+
+  def RemoveSshKey(self, key, md_key, level=None, store=True):
+    """Remove an ssh key to the metadata
+
+    Args:
+      key: string, the key to be removed.
+      md_key: string, SSH_KEYS or SSHKEYS_LEGACY, defines where to add the key.
+      level: enum, INSTANCE_LEVEL (default) or PROJECT_LEVEL to fetch the
+          metadata.
+      store: bool, if True, saves metadata to GCE server.
+    """
+    if not level:
+        level = self.INSTANCE_LEVEL
+    md_item = self.ExtractKeyItem(md_key, level)
+    md_item['value'] = re.sub('.*%s.*\n?' % key, '', md_item['value'])
+    if not md_item['value']:
+      self.md_items[level].remove(md_item)
+    if store:
+        self.StoreMetadata(level)
+
+  @RetryOnFailure
+  def TestSshLogin(self, key, expect_fail=False):
```

Comment:
This is fine for now, but you might want to consider splitting up utils based on functionality. It would probably make sense to have a `metadata_utils` and `instance_utils` (or something like that).